model.py

```diff
@@ -1,103 +1,1222 @@
+
+        pos: a dict saying where all nodes go if they have been assigned
+        parent: parent of this branch. - only affects it if non-directed
+
+        '''
+
+        if pos is None:
+            pos = {root: (xcenter, vert_loc)}
+        else:
+            pos[root] = (xcenter, vert_loc)
+
+        children = list(G.neighbors(root))
+
+        if not isinstance(G, nx.DiGraph) and parent is not None:
+            children.remove(parent)
+
+        if len(children) != 0:
+            dx = width/len(children)
+            nextx = xcenter - width/2 - dx/2
+            for child in children:
+                nextx += dx
+                pos = self._hierarchy_pos(
+                    G, child,
+                    width=dx, vert_gap=vert_gap,
+                    vert_loc=vert_loc-vert_gap, xcenter=nextx,
+                    pos=pos, parent=root
+                )
+
+        return pos
+
+
+# Subclasses.
+class Eclipse(Model):
+    '''Subclass of Model, specifically for storing a single eclipse. Lightcurve
+    data is stored on this level.
+
+    Inputs:
+    -------
+      lightcurve; Lightcurve:
+        A Lightcurve object, containing data
+      iscomplex; bool:
+        Do we use the complex model?
+      label; str:
+        A label to apply to the node. Mostly used when searching trees.
+      parameter_objects; list(Param), or Param:
+        The parameter objects that correspond to this node. Single Param is
+        also accepted.
+      parent; Model, optional:
+        The parent of this node.
+      children; list(Model), or Model:
+        The children of this node. Single Model is also accepted
+    '''
+
+    # Define this subclasses parameters
+    node_par_names = (
+        'dFlux', 'sFlux', 'ulimb', 'rdisc',
+        'scale', 'az', 'fis', 'dexp', 'phi0'
+    )
+
+    def __init__(self, lightcurve, iscomplex, *args, **kwargs):
+        # If we're a complex eclipse, add the complex parameter names
+        if iscomplex:
+            complex_parNames = (
+                'exp1', 'exp2', 'yaw', 'tilt'
+            )
+            self.node_par_names += complex_parNames
+
+        self.iscomplex = iscomplex
+
+        super().__init__(*args, **kwargs)
+
+        # If the lightcurve is a Lightcurve object, save it. Otherwise,
+        # read it in from the file.
+        if isinstance(lightcurve, Lightcurve):
+            self.lc = lightcurve
+        elif isinstance(lightcurve, str):
+            self.lc = Lightcurve.from_calib(lightcurve)
+        else:
+            msg = "Argument lightcurve is not a string or Lightcurve! "
+            msg += "Got {}".format(lightcurve)
+            raise TypeError(msg)
+
+        self.initCV()
+
+    def initCV(self):
+        '''Try to create a CV.
+        If we cant construct the input list of params, set cv to None.'''
+
+        self.cv = CV(self.cv_parlist)
+        # print("Created a CV object for eclipse {}...".format(self.label))
+
+    def calc(self, plot=False):
```

Comment:
yeah - I think i understand my confusion. To my mind, ```calc``` should return the result of the model, and a separate ```chisq``` should calculate chisq. 

I also think in general they should accept arguments for the data, rather than assuming a leaf node carries around it's own data, but I'm open to persuasion here.

Reply:
I think my above solution could sidestep this issue, as Eclipse subclasses could have their own data and call them in their version of evaluate_model(), but other use-cases could have their own stuff? Something like

```
def ln_like(self, *args, **kwargs):
    if self.is_leaf:
        self.evaluate_model(*args, **kwargs)
    ...

def evaluate_model(self, *args, **kwargs):
    data = kwargs['data']
    ...
```

Reply:
The Eclipse class now has a calcFlux() method that generates the model lightcurve. Now it looks like this:

- ln_like calls evaluate_model()
- evaluate_model() calls calcFlux() for the lightcurve
- evaluate_model() works out chisq, returns -0.5*chisq
- ln_like() returns the log likelihood